core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java

```diff
@@ -316,6 +316,18 @@ public void writeStringArrayNullable(@Nullable String[] array) throws IOExceptio
         }
     }
 
+    /**
+     * Writes a string array, for nullable string, writes it as 0 (empty string).
+     */
```

Comment:
This exactly the reason. I first tried `writeStringArrayNullable`, but then you read back an empty array, not a null array.